### PR TITLE
issue#16: rsb added the coinex exchange

### PR DESCRIPTION
**Pull request by rsbst23**
This PR includes the changes to integrate the Coinex exchange into ccxws.
Coinex does not have multi-market support, so I also created the basic-multiclient class to handle these types of markets.

Issue#16

### CODE REVIEW

File: src/enums.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+var MarketObjectTypes = Object.freeze({
```
Comment (bmancini55) [inline src/enums.js]: super minor: const instead of var.  this is very very important /s

File: src/exchanges/coinex-client.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    );
+  }
+
+  _onMessage(raw) {
+    let msg = JSON.parse(raw);
+
+    let { method, params } = msg;
+
+    // if params is defined, then this is a response to an event that was subscribed to, otherwise this is the initial connection response
+    if (params) {
```
Comment (bmancini55) [inline src/exchanges/coinex-client.js]: minor code structure thing, but may be cleaner to remove indent for the whole block by just doing 
```
if(!params) return;
```

File: src/exchanges/coinex-client.spec.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    expect(typeof ticker.ask).toBe("undefined");
+    expect(typeof ticker.askVolume).toBe("undefined");
+    expect(parseFloat(ticker.last)).toBeGreaterThan(0);
+    expect(parseFloat(ticker.open)).toBeGreaterThan(0);
+    expect(parseFloat(ticker.high)).toBeGreaterThan(0);
+    expect(parseFloat(ticker.low)).toBeGreaterThan(0);
+    expect(parseFloat(ticker.volume)).toBeGreaterThan(0);
+    expect(parseFloat(ticker.quoteVolume)).toBeGreaterThan(0);
+    expect(Math.abs(parseFloat(ticker.change))).toBeGreaterThan(0);
+    expect(Math.abs(parseFloat(ticker.changePercent))).toBeGreaterThan(0);
```
Comment (bmancini55) [inline src/exchanges/coinex-client.spec.js]: Minor change:  This is actually a bug in other tests and a cause of a lot of failed tests. change and  changePercent should do a check that it is not NaN.  I think this should be .not.isFalsey().  I'm adding an issue to fix this in other files as well.

File: src/basic-multiclient.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+  close(emitClosed = true) {
+    this._clients.forEach(c => {
+      c.close();
+    });
+
+    if (emitClosed) this.emit("closed");
+  }
+
+  _subscribe(market, map, marketObjectType, msg) {
```
Comment (bmancini55) [inline src/basic-multiclient.js]: Perfect implementation of event propagation in this function.  

My only concern is that it looks like it's reusing the client on multiple "types" of subscriptions for the same market.  Is this allowed for Coinex?  I was thinking we would create a client for each type and each market.  Basically create a map for each type and create a new client for each type+market.  Thoughts?

Comment (rsbst23) [reply src/basic-multiclient.js]: Oh, I thought that would be allowed. I thought it just wasn't allowed to reuse the client for the same type. 

I'm pretty sure I ran some tests with all 3 subscription types and got results in the command line. I should add a test around this to confirm it is working as expected and that it continues to deliver trades after i received other types of responses for example. I will double check that. If that doesn't work, then I will change it to do what you said.

Comment (bmancini55) [reply src/basic-multiclient.js]: Cool sounds good.  And if we have a use case down the line where it has to be one call per sub, then we can easily modify `_subscribe` to accommodate both.  The code coverage around that conditions will ensure Coinex continues to function as expected.

File: src/exchanges/coinex-client.spec.js (hunk 3 above)
Comment (bmancini55) [reply src/exchanges/coinex-client.spec.js]: BTW add issue #23 for fixing this in other tests.

File: src/basic-multiclient.js (hunk 4 above)
Comment (rsbst23) [reply src/basic-multiclient.js]: This was behaving as I thought it was. I did add a test around it, so like you said, if we have to refactor basic-multiclient.js to support other multi-markets in the future that don't behave like this, then we easily can.

**Pull request merged**